### DISCUSSION TITLE
MSP message to get Compass/Magnetomer data

### DISCUSSION THREAD

**Discussion by willishf** [iNavFlight/inav#8991] [Q&A]
Installed a standalone magnetometer to test with GPS-RTK ublox that does not have a magnetometer built in. 

Sourced the sensor from amazon https://www.amazon.com/gp/product/B0855TZV1J and is listed as a HMC5883L but based on discussion/feedback from amazon that board now appears to have QMC5883L which has a different I2C address and potentially different code requirements for setup/calibration etc. 

In iNAV setup selected QMC5883L and was found so that part is good. 

Went looking for the MSP command to pull the specific pitch, roll, yaw from this device and started with MSP_ATTITUDE_DATA. Not clear if this is from the accelerometer or fusion data of the accelerometer/compass. In testing the yaw is very slow to update when doing a 90 degree turn by hand on table but doesn't seem to come close to 90 degrees. I need to calibrate in that drone is currently taken apart for wiring so not easy to do the calibration. 

Also not clear what units the yaw value is getting returned. Getting values 20,000 and up so if degrees need to divide by 100. I didn't see this in any code samples. 

From looking at fc_msp.c it appears that MSP2_SENSOR_COMPASS is for receiving external compass data. 

Any guidance on what command to use to read any configured Compass/Magnetometer? 

We have a downward-looking camera and based on GPS position when taking a picture need accurate/absolute yaw information so that the image can be rotated.

**Comment by DzikuVx** [comment]
`MSP_RAW_IMU` has all the data you need
